### ISSUE TITLE
Non-breaking space problem

### ISSUE THREAD

**Issue by isaomatsunami** [OpenRefine/OpenRefine#4425]
Recently I encountered this problem on OpenRefine.
(Python code is just for explanation)

txt1 = "Ms. Wendy Enzmann"
txt2 = "Ms. Wendy Enzmann"
print( txt1 == txt2 ).    ===> False

The answer is;

print( ord(txt1[3]), ord(txt2[3]) ) ===> 32 160

32 is "space" in Ascii code. 160 is no-break space(NBSP) which is used to prevent automatic line break.

According to Wikipedia(https://en.wikipedia.org/wiki/Non-breaking_space), several non-breaking variants are defined in Unicode:

U+202F   NARROW NO-BREAK SPACE (HTML &#8239; · NNBSP)
U+2007   FIGURE SPACE (HTML &#8199; · &numsp;)
U+2060 WORD JOINER (HTML &#8288; · &NoBreak; · WJ)

If a string contains NBSPs like txt2, toTitlecase() function and facet/cluster features in OpenRefine behave like a bug. 

For the time being, value.replace(" "," "), replacing NBSP wih SPACE works effectively. But this is not for general OpenRefine users.

What should we do?

**Comment by wetneb** [comment]
Hello Isao, nice to hear from you!

Two things come to mind:
* perhaps it would make sense to have an operation to replace non-standard whitespace by space characters (for instance in the Common transforms menu). It is something you would have to trigger yourself ahead of other transforms (such as titleCase).
* it would be nice if OpenRefine displayed those special characters in… a special way! In fact we already have an issue about displaying non-printable characters, it would make sense to extend this for non-standard whitespace, perhaps? https://github.com/OpenRefine/OpenRefine/issues/1286

What do you think?

**Comment by thadguidry** [comment]
Hi Isao and Antonin,

We would also need to make decisions on inclusion/exclusion of "combining
space mark" and maybe allowing users to control how to treat them.  For
example, including the characters of the "Mc" class additionally to the
"Zs" space separator class (or not).  I do not think most users will want
to do that but instead have some ability to know where any "Mc" or "Zs"
class is located and be able to replace them.
I think that is currently possible if you utilize our existing
unicodeType() function with a replace() ?

We already have support in the unicodeType() function class:

https://github.com/OpenRefine/OpenRefine/blob/master/main/src/com/google/refine/expr/functions/strings/UnicodeType.java#L66
https://github.com/OpenRefine/OpenRefine/blob/master/main/src/com/google/refine/expr/functions/strings/UnicodeType.java#L70

Codepoint 8 is the abstraction what covers the "Mc" class in Unicode that
maps to the COMBINING_SPACING_MARK
<https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Character.html#COMBINING_SPACING_MARK>
as
shown in
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Character.html
And it has various definitions around:

https://docs.microsoft.com/en-us/dotnet/api/system.globalization.unicodecategory?view=net-6.0

> SpacingCombiningMark -
>
> Spacing character that indicates modifications of a base character and
> affects the width of the glyph for that base character. Signified by the
> Unicode designation "Mc" (mark, spacing combining). The value is 6.
>
>
Also there are about 446 such characters in the "Mc" class of of Unicode
14.0 if you search this doc with ";Mc;"
https://www.unicode.org/Public/UCD/latest/ucd/UnicodeData.txt

And here's the current Unicode - General Category reference:
https://www.unicode.org/reports/tr44/#General_Category_Values

Thad
https://www.linkedin.com/in/thadguidry/
https://calendly.com/thadguidry/

**Comment by isaomatsunami** [comment]
In my opinion, Most of OpenRefine users use it, with real-world entities in mind, to process NOT data bytes BUT concrete concepts. So non-discernible bytes should not have any effects. Who among librarians/journalist/academics needs to have the knowledge of NBSP or other unprintable bytes? OpenRefine already uses Unicode normalization(NFKC) conversion.

Presumably you want to keep these bytes. Yes, you are dev-nerd in nature !! But I think these bytes should be replaced or deleted, as default, when imported, but this default setting can be opted-out for knowledgeable users. 

This is just my opinion.

**Comment by thadguidry** [comment]
@isaomatsunami Not all users have the same needs, many are data journalists and scientists and the options provided need to be thought through.  For example, the original source data needs some opinionated cleaning and byte replacement by inspection from human eyes.  We want to give users ways to 1. quickly visualize the category of data problems with spaces and then 2. offering better transformations.  Which are basically the 2 bullet points that @wetneb mentions.

**Comment by wetneb** [comment]
Yes, but it is also true that we do already offer *some* normalization options at import time, such as trimming whitespace around cells, so it would not be completely unimaginable to offer more of those features there, right?

That being said, one problem I see with normalizing this at import time is that it does not cover data that gets into the project afterwards (fetching URLs, data extension, reconciliation, and even just the result of evaluating expressions which can always insert new gibberish things). So if we really wanted to make sure those characters never end up in a cell, we would need to have fairly aggressive normalization procedures in many places in the code (or at least they would be run very often).

**Comment by isaomatsunami** [comment]
You both know OpenRefine much much better. I leave this entirely to you.

I just wanted to report what I encountered when I merged plain text data and scraped data. This will be a rare case, prompted by environment changes caused, of cause,  by this pandemic!
I had nearly posted a bug report.

**Comment by thadguidry** [comment]
> Yes, but it is also true that we do already offer _some_ normalization options at import time, such as trimming whitespace around cells, so it would not be completely unimaginable to offer more of those features there, right?

@wetneb Right.

**Comment by tfmorris** [comment]
In places where the space character is an ancient, turn-of-the-millennium, proxy for a word separator it would be worth considering the more modern Unicode guidelines for identifying word boundaries: https://www.unicode.org/reports/tr29/tr29-37.html#Word_Boundaries

**Comment by thadguidry** [comment]
1. I think this is one of those issues that is a surprise to users because things are hidden, and no warning is given to a user about hidden characters.  At a minimum it seems we should offer a warning dialog, after the project is created to notify that some cells or rows contain hidden characters.
2. Then what suggestion to view or help remedy would the dialog suggest? The unicode facet isn't available under the All menu column, but something like it that focuses only on showing rows with cells that contain hidden characters might be a good start?
